### PR TITLE
[view-transitions] Implement DOM API necessary to support :active-view-transition-type()

### PR DESCRIPTION
**Pull request by tuankiet65**
#### c6cadba6b01325cd2e8a3b35d8f0ef440ea97429
<pre>
[view-transitions] Implement DOM API necessary to support :active-view-transition-type()
<a href="https://bugs.webkit.org/show_bug.cgi?id=277777">https://bugs.webkit.org/show_bug.cgi?id=277777</a>
<a href="https://rdar.apple.com/133419258">rdar://133419258</a>

Reviewed by Tim Nguyen.

This commit sets up the necessary infrastructure on the JavaScript DOM
API side to support :active-view-transition-type(). This includes:
* Adding a new setting called ViewTransitionsTypeEnabled
* Extending Document.startViewTransition to additionally accept
  StartViewTransitionOptions
* Implementing the ViewTransitionTypeSet object
* Extending ViewTransition to include a `types` member of type
  ViewTransitionTypeSet.

* LayoutTests/TestExpectations: Removed one passing test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/navigation/with-types/types-in-pagereveal-and-pageswap-expected.txt:
Rebaselined.
* LayoutTests/imported/w3c/web-platform-tests/css/css-view-transitions/view-transition-types-mutable-expected.txt:
Rebaselined now that the test is passing.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml: Added new setting ViewTransitionsTypeEnabled
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document+ViewTransition.idl: Extended Document.startViewTransition
to additionally accept StartViewTransitionOptions.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::startViewTransition): Extended to accept StartViewTransitionOptions
and pass the types within the options to ViewTransition.
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/StartViewTransitionOptions.h: Added.
* Source/WebCore/dom/StartViewTransitionOptions.idl: Added.
* Source/WebCore/dom/ViewTransition+Types.idl: Added new member `types` of type
ViewTransitionTypeSet.
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::ViewTransition): Additionally accept a vector of types and pass it
to ViewTransitionTypeSet.
(WebCore::ViewTransition::createSamePage):
* Source/WebCore/dom/ViewTransition.h:
(WebCore::ViewTransition::types const):
(WebCore::ViewTransition::setTypes):
* Source/WebCore/dom/ViewTransitionTypeSet.cpp: Added.
(WebCore::ViewTransitionTypeSet::ViewTransitionTypeSet):
(WebCore::ViewTransitionTypeSet::initializeSetLike const):
(WebCore::ViewTransitionTypeSet::clearFromSetLike):
(WebCore::ViewTransitionTypeSet::addToSetLike):
(WebCore::ViewTransitionTypeSet::removeFromSetLike):
(WebCore::ViewTransitionTypeSet::has const):
* Source/WebCore/dom/ViewTransitionTypeSet.h: Added.
(WebCore::ViewTransitionTypeSet::create):
* Source/WebCore/dom/ViewTransitionTypeSet.idl: Added.

Canonical link: <a href="https://commits.webkit.org/282088@main">https://commits.webkit.org/282088@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/78095386d6d36d15c241feee50cf953d5430495c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62013 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41367 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14605 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65993 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12558 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64132 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49053 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12834 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49935 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8661 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65082 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53700 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30767 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35056 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10936 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11489 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/55108 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56834 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11240 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67721 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/61255 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5956 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10942 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57312 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5982 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53649 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57559 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13788 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4875 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/83018 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37167 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14548 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38251 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39347 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37996 "Built successfully") | | |

### CODE REVIEW

File: Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml [hunk 1]
```diff
@@ -7420,6 +7420,20 @@ ViewTransitionsEnabled:
     WebCore:
       default: true
 
+ViewTransitionsTypeEnabled:
```
Comment (nt1m) [inline Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml]:
```suggestion
ViewTransitionTypesEnabled:
```

File: Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml [hunk 2]
```diff
@@ -7420,6 +7420,20 @@ ViewTransitionsEnabled:
...
+  type: bool
+  category: animation
+  status: testable
+  humanReadableName: "View Transitions :active-view-transition-type() support"
+  humanReadableDescription: "Support specifying active view transition types"
```
Comment (nt1m) [inline Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml]:
```suggestion
  humanReadableName: "View Transition Types"
  humanReadableDescription: "Support specifying types for view transitions"
```

It's not just about the CSS selector fwiw.

File: Source/WebCore/dom/Document.cpp [hunk 3]
```diff
@@ -10723,19 +10725,51 @@ void Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges()
...
 
-RefPtr<ViewTransition> Document::startViewTransition(RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
+ExceptionOr<RefPtr<ViewTransition>> Document::startViewTransition(StartViewTransitionCallbackOptions&& callbackOptions)
 {
     if (!globalObject())
         return nullptr;
 
-    Ref viewTransition = ViewTransition::create(*this, WTFMove(updateCallback));
+    if (callbackOptions && std::holds_alternative<StartViewTransitionOptions>(*callbackOptions) && !settings().viewTransitionsTypeEnabled())
+        return Exception { ExceptionCode::TypeError };
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]: I think it's probably OK to remove this and let people do `{ update: () => {} }` even if the types preference is disabled. The types stuff is properly gated in the IDL.

File: Source/WebCore/dom/Document.cpp [hunk 4]
```diff
@@ -10723,19 +10725,51 @@ void Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges()
...
+    RefPtr<ViewTransitionUpdateCallback> updateCallback = nullptr;
+
+    Vector<AtomString> activeTypes { };
+
+    // 2. If callbackOptions is an an UpdateCallback, set updateCallback to callbackOptions.
+    // 3. Otherwise, if callbackOptions is a StartViewTransitionOptions, then set updateCallback to callbackOptions’s update.
+    // Also extract active types for step 6.
+    if (callbackOptions) {
+        WTF::switchOn(*callbackOptions, [&](const RefPtr<JSViewTransitionUpdateCallback>& argUpdateCallback) {
+            updateCallback = WTFMove(argUpdateCallback);
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]:
```suggestion
        WTF::switchOn(*callbackOptions, [&](const RefPtr<JSViewTransitionUpdateCallback>& callback) {
            updateCallback = WTFMove(callback);
```

File: Source/WebCore/dom/Document.cpp [hunk 5]
```diff
@@ -10723,19 +10725,51 @@ void Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges()
...
+        }, [&](const StartViewTransitionOptions& transitionOptions) {
+            ASSERT(settings().viewTransitionsTypeEnabled());
+
+            updateCallback = WTFMove(transitionOptions.update);
+            if (transitionOptions.types)
+                activeTypes = WTFMove(*transitionOptions.types);
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]:
```suggestion
        }, [&](const StartViewTransitionOptions& options) {
            ASSERT(settings().viewTransitionsTypeEnabled());

            updateCallback = WTFMove(options.update);
            if (options.types)
                activeTypes = WTFMove(*options.types);
```

File: Source/WebCore/dom/Document.cpp [hunk 6]
```diff
@@ -10723,19 +10725,51 @@ void Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges()
     }
 }
 
-RefPtr<ViewTransition> Document::startViewTransition(RefPtr<ViewTransitionUpdateCallback>&& updateCallback)
+ExceptionOr<RefPtr<ViewTransition>> Document::startViewTransition(StartViewTransitionCallbackOptions&& callbackOptions)
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]: Let's not make this throw an exception and just accept both forms of the argument.

File: Source/WebCore/dom/Document.cpp [hunk 7]
```diff
@@ -10723,19 +10725,51 @@ void Document::flushDeferredRenderingIsSuppressedForViewTransitionChanges()
...
+    Ref viewTransition = ViewTransition::create(*this, WTFMove(updateCallback), WTFMove(activeTypes));
 
     if (RefPtr activeViewTransition = m_activeViewTransition)
         activeViewTransition->skipViewTransition(Exception { ExceptionCode::AbortError, "Old view transition aborted by new view transition."_s });
 
     setActiveViewTransition(WTFMove(viewTransition));
     scheduleRenderingUpdate(RenderingUpdateStep::PerformPendingViewTransitions);
-    return m_activeViewTransition;
+
+    return { m_activeViewTransition.copyRef() };
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]: Why this change?

**Review by nt1m — COMMENTED** [review]
Looks OK, I mostly have nits. Please fix the GTK/WPE build

File: Source/WebCore/dom/Document.cpp (hunk 7 above)
Comment (tuankiet65) [reply Source/WebCore/dom/Document.cpp]: It's needed to return `ExceptionOr<RefPtr<...>>` (or I'm wrong and there's a different way). Anyway, now that we don't need to raise an exception, this isn't needed anymore.

File: Source/WebCore/dom/ViewTransition.h [hunk 8]
```diff
@@ -159,10 +160,13 @@ class ViewTransition : public RefCounted<ViewTransition>, public CanMakeWeakPtr<
 
     bool documentElementIsCaptured() const;
 
+    const ViewTransitionTypeSet& types() const { return m_types; }
+    void setTypes(Ref<ViewTransitionTypeSet>&& newTypes) { m_types = newTypes; }
```
Comment (nt1m) [inline Source/WebCore/dom/ViewTransition.h:185]:
```suggestion
    void setTypes(Ref<ViewTransitionTypeSet>&& newTypes) { m_types = WTFMove(newTypes); }
```

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/dom/ViewTransition.h (hunk 8 above)
Comment (tuankiet65) [reply Source/WebCore/dom/ViewTransition.h:185]: I forgot that you need to explicitly use `WTFMove` in this scenario 😅

File: Source/WebCore/dom/Document.h [hunk 9]
```diff
@@ -404,6 +406,8 @@ using RenderingContext = std::variant<
     RefPtr<CanvasRenderingContext2D>
 >;
 
+using StartViewTransitionCallbackOptions = std::optional<std::variant<WTF::RefPtr<JSViewTransitionUpdateCallback>, StartViewTransitionOptions>>;
```
Comment (nt1m) [inline Source/WebCore/dom/Document.h]: Is it possible to eliminate the `std::optional` here? RefPtr should already be a nullable type.

Also, `WTF::RefPtr` can probably be `RefPtr`.

Comment (nt1m) [reply Source/WebCore/dom/Document.h]: Alternatively, I'd also be OK with changing `RefPtr` to `Ref` and keeping the optional.

Comment (tuankiet65) [reply Source/WebCore/dom/Document.h]: The parameter type in IDL is

```
optional (ViewTransitionUpdateCallback or StartViewTransitionOptions) callbackOptions = {}
```

This should map to a `std::optional<std::variant<..., ...>>` right? If we change it to `std::variant<WTF::RefPtr<JSViewTransitionUpdateCallback>, StartViewTransitionOptions>` then not specifying a parameter is equivalent to setting it to `nullptr` which is a `WTF::RefPtr<JSViewTransitionUpdateCallback>`?

Comment (nt1m) [reply Source/WebCore/dom/Document.h]: Ref instead of RefPtr would be more accurate then:

`std::optional<std::variant<Ref<JSViewTransitionUpdateCallback>, StartViewTransitionOptions>`

`RefPtr` would correspond to `ViewTransitionUpdateCallback?` rather than `ViewTransitionUpdateCallback`

Comment (tuankiet65) [reply Source/WebCore/dom/Document.h]:
```
/Volumes/Data/Sources/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/JSDocument.cpp:6787:175: 
error: no viable conversion from 'optional<variant<WTF::RefPtr<WebCore::JSViewTransitionUpdateCallback>, [...]>>' to 'optional<variant<Ref<JSViewTransitionUpdateCallback>, [...]>>'
```

Oof, seems like the IDL compiler still compiles `ViewTransitionUpdateCallback` to `RefPtr<JSViewTransitionUpdateCallback>`

Comment (nt1m) [reply Source/WebCore/dom/Document.h]: OK let's try to land as is then. But please fix the GTK/WPE build first

**Comment by tuankiet65** [comment]
There are some additions I haven't added (like adding `WTFMove` to `setType`) 😅 I'll get them into the next patch then
